web/src/engine/osk/src/keyboard-layout/oskKey.ts

```diff
@@ -310,47 +294,73 @@ export default abstract class OSKKey {
     }
 
     // Check the key's display width - does the key visualize well?
-    let emScale = vkbd.getKeyEmFontSize();
-    var width: number = getTextMetrics(keyText, emScale, styleSpec).width;
-    if(width == 0 && keyText != '' && keyText != '\xa0') {
-      // Add the Unicode 'empty circle' as a base support for needy diacritics.
-
-      // Disabled by mcdurdin 2020-10-19; dotted circle display is inconsistent on iOS/Safari
-      // at least and doesn't combine with diacritic marks. For consistent display, it may be
-      // necessary to build a custom font that does not depend on renderer choices for base
-      // mark display -- e.g. create marks with custom base included, potentially even on PUA
-      // code points and use those in rendering the OSK. See #3039 for more details.
-      // keyText = '\u25cc' + keyText;
-
-      if(vkbd.isRTL) {
-        // Add the RTL marker to ensure it displays properly.
-        keyText = '\u200f' + keyText;
-      }
+    if(vkbd.isRTL) {
+      // Add the RTL marker to ensure it displays properly.
+      keyText = '\u200f' + keyText;
     }
 
-    ts.fontSize = this.getIdealFontSize(vkbd, keyText, styleSpec);
-
     // Finalize the key's text.
     t.innerText = keyText;
 
     return t;
   }
 
-  public refreshLayout(vkbd: VisualKeyboard) {
+  public resetFontPrecalc() {
+    this._fontFamily = undefined;
+    this._fontSize = undefined;
+    this.label.style.fontSize = '';
+  }
+
+  public refreshLayout(layoutParams: KeyLayoutParams) {
+    // Avoid doing any font-size related calculations if there's no text to display.
+    if(this.spec.sp == ButtonClasses.spacer || this.spec.sp == ButtonClasses.blank) {
+      return () => {};
```

Comment:
It'd be better to `return null` and then in the caller, use `closure?.()` to make the call: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Optional_chaining#optional_chaining_with_function_calls